clash-cores/test/Test/Cores/Xilinx/DcFifo.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Test.Cores.Xilinx.DcFifo ( tests ) where
+
+import qualified Prelude as P
+import qualified Data.List as L
+import Data.Maybe (catMaybes)
+import Control.Monad (replicateM)
+
+import Clash.Explicit.Prelude
+import Clash.Cores.Xilinx.DcFifo.Explicit
+
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Hedgehog (MonadGen, (===), property, forAll, Property)
+import Test.Tasty.Hedgehog (testPropertyNamed)
+import Test.Tasty (TestTree)
+
+tests :: TestTree
+tests = testPropertyNamed
+  "FIFO doesn't lose any data"
+  "prop_fifo"
+  prop_fifo
+
+intersperseStalls ::
+  -- | Data
+  [a] ->
+  -- | Stalls
+  [a] ->
+  [a]
+intersperseStalls (d:ds) s = d:s P.++ intersperseStalls ds s
+intersperseStalls [] _ = []
+
+prop_fifo :: Property
+prop_fifo = property $ do
+  xs <- fmap Just <$> forAll (replicateM 10 genData)
+  stallRead <- forAll (Gen.maybe (Gen.int (Range.linear 0 10)))
+  stallWrite <- forAll (Gen.maybe (Gen.int (Range.linear 0 10)))
```

Comment:
To maximize the chaos, you want to generate one int per driven element. By doing this you generate `n` cycles of stall for every element driven. Most programming errors occur around things like bursts of data, followed by nothing, etc.

Reply:
It's probably easiest to zip these generated ints with the elements you're driving. In that case you can simply count down in your drive/sample functions.

Reply:
Right! I have variable stalls now.